plugins/serialization/src/main/java/org/qbicc/plugin/serialization/HeapSerializer.java

```diff
@@ -1,426 +1,44 @@
-        Value array = bb.newArray(jlo.getType().getReferenceArrayObject(), original.getParameterValue(1)); // Can't elide the clear inside of new; might trigger a GC when deserializing elements
-        ValueHandle arrayHandle = bb.referenceHandle(array);
-        bb.store(bb.instanceFieldOf(arrayHandle, layout.getRefArrayElementTypeIdField()), original.getParameterValue(0), MemoryAtomicityMode.UNORDERED);
-        bb.return_(array);
-
-        bb.finish();
-        meth.replaceMethodBody(MethodBody.of(bb.getFirstBlock(), Schedule.forMethod(bb.getFirstBlock()), original.getThisValue(), original.getParameterValues()));
-    }
-
-
-    // Implementation of RuntimeObjectDeserializer.createString(byte[], byte, bool, bool)
-    private static void createStringImpl(CompilationContext ctxt, MethodElement meth) {
-        BasicBlockBuilder bb = ctxt.getBootstrapClassContext().newBasicBlockBuilder(meth);
-        MethodBody original = meth.getMethodBody();
-
-        bb.begin(new BlockLabel());
-        LoadedTypeDefinition jls = ctxt.getBootstrapClassContext().findDefinedType("java/lang/String").load();
-        Value str = bb.new_(jls.getClassType());
-        bb.store(bb.instanceFieldOf(bb.referenceHandle(str), jls.findField("value")), original.getParameterValue(0), MemoryAtomicityMode.UNORDERED);
-        bb.store(bb.instanceFieldOf(bb.referenceHandle(str), jls.findField("coder")), original.getParameterValue(1), MemoryAtomicityMode.UNORDERED);
-        bb.return_(str);
-
-        bb.finish();
-        meth.replaceMethodBody(MethodBody.of(bb.getFirstBlock(), Schedule.forMethod(bb.getFirstBlock()), original.getThisValue(), original.getParameterValues()));
-    }
-
-    // Implementation of RuntimeObjectDeserializer.deserializeInstanceFields(int, Object, Deserializer)
-    private static void deserializeInstanceFieldsImpl(CompilationContext ctxt, MethodElement meth, LoadedTypeDefinition[] classes) {
-        BasicBlockBuilder bb = ctxt.getBootstrapClassContext().newBasicBlockBuilder(meth);
-        MethodBody original = meth.getMethodBody();
-        LoadedTypeDefinition deser = ctxt.getBootstrapClassContext().findDefinedType("org/qbicc/runtime/deserialization/Deserializer").load();
-
-        int numCases = classes.length;
-        int[] typeIds = new int[numCases];
-        BlockLabel[] targets = new BlockLabel[numCases];
-        BlockLabel errorBlock = new BlockLabel();
-        for (int i=0; i<numCases; i++) {
-            targets[i] = new BlockLabel();
-            typeIds[i] = classes[i].getTypeId();
-        }
-
-        bb.begin(new BlockLabel());
-        bb.switch_(original.getParameterValue(0), typeIds, targets, errorBlock);
+        LoadedTypeDefinition ih = ctxt.getBootstrapClassContext().findDefinedType("org/qbicc/runtime/main/InitialHeap").load();
+        Section heapSection = ctxt.getOrAddProgramModule(ih).getOrAddSection(ctxt.IMPLICIT_SECTION_NAME); // TODO: use ctxt.INITIAL_HEAP_SECTION_NAME
```

Comment:
If I try to use the INITIAL_HEAP_SECTION_NAME here, I get an error from clang backend:
```
fatal error: error in backend: Global variable 'qbicc_initial_heap_obj_0' has an invalid section specifier 'QBICC_I_HEAP': mach-o section specifier
      requires a segment and section separated by a comma.
```

Reply:
I think we can look into this independently of this PR.